### PR TITLE
[CALCITE-5619] Add support for TO_CHAR

### PR DESCRIPTION
**Pull request by tindzk**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 1]
```diff
@@ -863,6 +863,16 @@ static RelDataType deriveTypeSplit(SqlOperatorBinding operatorBinding,
           OperandTypes.STRING.or(OperandTypes.BINARY),
           SqlFunctionCategory.STRING);
 
+  /** The "TO_CHAR(timestamp, format)" function;
+   * converts {@code timestamp} to string according to the given {@code format}.
+   */
+  @LibraryOperator(libraries = {MYSQL, ORACLE, POSTGRESQL})
+  public static final SqlFunction TO_CHAR =
```
Comment (dssysolyatin) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:870]: PostgreSQL has 3 different to_char:
`to_char ( timestamp, text ) → text`
`to_char ( interval, text ) → text`
`to_char ( numeric_type, text ) → text`

You implemented only it for timestamp type. Can you specify it in jira ticket

File: babel/src/test/resources/sql/postgresql.iq [hunk 2]
```diff
@@ -57,3 +57,8 @@ SELECT * FROM sal_emp;
 NAME, PAY_BY_QUARTER, SCHEDULE
 Bill, [10000, 10000, 10000, 10000], [[meeting, lunch], [training, presentation]]
 !ok
+
+select to_char(timestamp '2022-06-03 12:15:48.678', 'YYYY-MM-DD HH24:MI:SS.MS TZ');
```
Comment (dssysolyatin) [inline babel/src/test/resources/sql/postgresql.iq:61]: Can you add more tests for `toChar(DataContext ctx, long timestamp, String pattern)` function in org.apache.calcite.test.SqlFunctionsTest ?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 3]
```diff
@@ -583,6 +584,7 @@ Builder populate2() {
       defineMethod(TIME, "time", NullPolicy.STRICT);
 
       // Datetime formatting methods
+      map.put(TO_CHAR, new ToCharImplementor());
```
Comment (dssysolyatin) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:587]: Use `defineMethod` instead of `ToChatImplementor`

**Review by dssysolyatin — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java
Comment (zoudan) [inline core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java:1]: Shall we add some tests in SqlOperatorTest?

File: core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java [hunk 4]
```diff
@@ -165,10 +173,17 @@ public enum FormatElementEnum implements FormatElement {
...
+      final Calendar calendar = Work.get().calendar;
+      calendar.setTime(date);
+      return String.format(Locale.ROOT, "%03d", calendar.get(Calendar.MILLISECOND));
+    }
+  },
   TZR("The time zone name") {
     @Override public String format(Date date) {
       // TODO: how to support timezones?
-      throw new UnsupportedOperationException();
+      return "";
```
Comment (zoudan) [inline core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:186]: why we return empty string for this case

Comment (tindzk) [reply core/src/main/java/org/apache/calcite/util/format/FormatElementEnum.java:186]: We do so in order to match PostgreSQL's output when formatting `timestamp`s: https://dbfiddle.uk/VkeR6jUV

**Comment by dssysolyatin** [comment]
LGTM 

If @julianhyde and @zoudan do not have any objection then I will merge it

**Comment by zoudan** [comment]
LGTM

**Pull request closed without merging**